pkg/agent/openflow/pipeline.go

```diff
@@ -41,38 +41,128 @@ import (
+	// - Choose a stage. All tables in a stage are intended to implement a functional module in pipelines. Existing stages
+	//   are defined in file pkg/agent/openflow/feature.go. If you want to add a new stage, please discuss with maintainers
+	//   or OVS pipeline developers of Antrea.
+	// - Choose a target pipeline. Existing pipelines are defined in file pkg/agent/openflow/feature.go. If you want to add a
+	//   new pipeline, please discuss with maintainers or OVS pipeline developers of Antrea.
+	// - Decide where to add the new table in the pipeline. The order table declaration decides the order of tables in the
+	//   stage. For example:
+	//     * If you want to add a table called `DummyTable` between `SpoofGuardTable` and `IPv6Table` in pipelineIP, then
+	//       the table should be declared after `SpoofGuardTable` and before `IPv6Table`:
+	//       ```go
+	//          SpoofGuardTable  = newTable("SpoofGuard", stageValidation, pipelineIP)
+	//          DummyTable       = newTable("Dummy", stageValidation, pipelineIP)
+	//          IPv6Table        = newTable("IPv6", stageValidation, pipelineIP)
+	//	        PipelineIPClassifierTable = newTable("IPClassifier", stageValidation, pipelineIP)
+	//       ```
+	//      * If you want to add a table called `DummyTable` just before `ARPResponderTable` in pipelineARP, then the table
+	//        should be declared before `ARPResponderTable`:
+	//       ```go
+	//          DummyTable        = newTable("Dummy", stageOutput, binding.PipelineARP)
+	//          ARPResponderTable = newTable("ARPResponder", stageOutput, binding.PipelineARP)
+	//       ```
+	//       * If you want to add a table called `DummyTable` just after `ConntrackStateTable` in pipelineARP, then the
+	//         table should be declared after `ConntrackStateTable`:
+	//       ```go
+	//          SNATConntrackTable  = newTable("SNATConntrackZone", stageConntrackState, pipelineIP)
+	//          ConntrackTable      = newTable("ConntrackZone", stageConntrackState, pipelineIP)
+	//          ConntrackStateTable = newTable("ConntrackState", stageConntrackState, pipelineIP)
+	//          DummyTable          = newTable("Dummy", stageConntrackState, pipelineIP)
+	//       ```
+	//  - Reference the new table in a feature in file pkg/agent/openflow/feature.go. The table can be referenced by multiple
+	//    features if multiple features need to install flows in the table. Note that, if the newly added table is not
+	//    referenced by any feature or the features referencing the table are all inactivated, then the table will not
+	//    be realized in OVS; if at least one feature referencing the table is activated, then the table will be realized
+	//    at the desired position in OVS pipeline.
+	//  - For pipeline modularity, action GotoTable is no longer recommended. For packet forwarding within a stage, action
+	//    ResubmitToTables is recommended; for packet forwarding between stages, action GotoStage is recommended.
```

Comment:
In Flexible pipeline, we expect that:

- Packet forwarding between different stages should use GotoStage.
- Packet forwarding within a stage should use NextTable() as possible, if NextTable() cannot satisfy some cases, use resubmit instead.

Reply:
Perhaps my question is why we expect to use resubmit for tables within a stage, any pros of resubmit, any cons of goto. I have two arguments why it should use goto:
1. Previously it uses goto, no reason to make a change without a reason.
2. With goto, it's possible to achieve #326 in the future.

Reply:
Within a stage, we have to use resubmit since we have some existing related flows. In Flexible pipeline, I think we should just use a unified action (resubmit) within a stage.

Reply:
I think you mean the flows for AntreaProxy. We are aware of it when switching to GoTo, and I remember it doesn't affect #326 if partial flows use resubmit. But I still don't see how it's unified: NextTable() still uses GoTo, how does it look more unified when two cases (go to a stage and go to next table) use GoTo and one case (go to specific table in same stage) use resubmit. From openflows' perspective, doesn't it look more confusing that some actions use goto while some use resubmit, where this is no concept of stage?

I think it's easier to understand If we stick to goto for all cases except when it has to use resubmit for some scenarios?

cc @jianjuns @antoninbas @wenyingd for opinions.

Reply:
I agree we should generally recommend goto, and use resubmit only when it is a must.

Reply:
For a single OpenFlow entry, we could add multiple "resubmit" to table in the actions, but there is only one goto_table in the action. This is the difference between these two actions. 

From flexible pipeline aspect, we don't recommend a flow to goto any table in a different stage (we can't completely forbid this action if it is a must for some special cases, so we don't remove the API in pkg/ovs/openflow), since it might break the packet pipeline. But for the scenario a packet must skip some following tables inside the same stage, a flow to forward packet to another table is allowed, e.g., we have tables to process hairpin packets in Routing stage, but for the packet with a remote Pod CIDR destination these tables should be skip.

From my understanding, the comments here only means the API `GotoTable` (not referring to OpenFlow goto_table action) is not recommended to developers to avoid cross stages case. But for the case that packets forwarding between tables in the same stage,  API `GotoTable` is acceptable.

Reply:
I added the following comments and changed `ResubmitToTables` back to `GotoTable` in pipeline.go.

```
	//  - Next is used to forward packets to the next table. E.g, A -> B, B -> C, C -> D, etc.
	//  - GotoTable is used to forward packets to a specific table, and the target table ID should be greater than the
	//    current table ID. Note that, it is recommended to use GotoTable to forward packets with a stage, e.g, B -> D,
	//    C -> E. However, in future OVS pipeline development, if packets need to be forwarded to a table in another stage
	//    directly anyway, e.g, A -> C, B -> G, GoTable can also be used.
	//  - GotoStage is used to forward packets to a specific stage. Note that, packets are forwarded to the first table of
	//    the target stage, and the first table ID of the target stage should be greater than the current table ID. E.g,
	//    A -> S4 (F), D -> S4 (F) are fine, but D -> S1 (A), F -> S2 (B) are not allowed. It is recommended to use
	//    GotoStage to forward packets across stages.
	//  - ResubmitToTables is used to forward packets to one or multiple tables. E.g, B -> D E, E -> B. Note that, the
	//    target table ID can be smaller than current table ID, like E -> B.
```